xwayland/xwm.c

```diff
@@ -1215,7 +1233,21 @@ static void xwm_handle_client_message(struct wlr_xwm *xwm,
 		xwm_handle_wm_protocols_message(xwm, ev);
 	} else if (ev->type == xwm->atoms[NET_ACTIVE_WINDOW]) {
 		xwm_handle_net_active_window_message(xwm, ev);
-	} else if (!xwm_handle_selection_client_message(xwm, ev)) {
+	} else if (ev->type == xwm->atoms[WM_CHANGE_STATE]) {
+		struct wlr_xwayland_surface *xsurface = lookup_surface(xwm, ev->window);
+		uint32_t detail = ev->data.data32[0];
+
+		if (xsurface == NULL) {
+			return;
+		} else if (detail == ICCCM_ICONIC_STATE) {
```

Comment:
What is the relationship between `ICCCM_ICONIC_STATE` and `NET_WM_STATE_HIDDEN`?

Reply:
if a window is minimized it's both the NET_WM_STATE is hidden the WM_STATE is iconic

Reply:
Hm, so why do we need to handle `ICCCM_ICONIC_STATE` changes?

Reply:
Bump

Reply:
This is the way minimize is being set - through WM_STATE which is iconic state - i3 does the same, net_wm_state_hidden mean window is not visible, iconic state means minimized

Reply:
Additionally you can do the following to test what clients do,
run you compositor with debug, so WLR_DEBUG is printed.
Click minimize button on chrome or steam csd, you will see WM_CHANGE_STATE
event. The WM_STATE has 3 types (minimized - iconic - 3) (normal - normal - 1)
and withdrawn which is basically closed.

Reply:
Also you can compare the status of steam for example in minimized and unminimized 
state, both things WM_STATE and _NET_WM_STATE are changed.
The change signal / request comes via WM_STATE.

Reply:
Please refrain from marking open discussions as resolved.

Okay, so I guess WM_CHANGE_STATE with ICCCM_ICONIC_STATE is a way for clients to request to be minimized. In this case, this should go through the `request_minimize` event.

Reply:
Not sure what your above statement means code-wise, 
or does it just mean 'resolved'.

Reply:
There are two code-paths in this PR:

- Client wants to become minimized. It sends `NET_WM_STATE_HIDDEN`, and if I understand correctly, it can alternatively send `ICCCM_ICONIC_STATE`. In this case, wlroots should (1) not minimize the window and (2) emit a `request_minimize` event. (1) is important because we need to let the compositor handle the request. Maybe the compositor can't minimize windows. Maybe the compositor wants to do something else when a client requests to be minimized. The compositor must stay in control. wlroots shouldn't do anything behind the compositor's back.
- A compositor decides to minimize a window. This can be in response to a `request_minimize` event, or this can be because of something else (e.g. the user clicked "minimize" in a compositor menubar). The compositor will call `wlr_xwayland_surface_set_minimized`.

Reply:
@damianatorrpm this discussion needs to be resolved - this is the main blocker for this patch.

wlroots isn't a window manager in its own right, its a toolkit for window managers. We just help information flow smoothly from client to compositor implementation, and the compositor implementation calls the shots. So, instead of having wlroots' Xwayland layer handle the minimize request, we need to add a new event which the compositor can register for, which is emitted when the window requests to be minimized. Then, the compositor can decide what to do - for example, calling `wlr_xwayland_surface_set_minimized` themselves.

Reply:
This has been changed, see the updated PR.

Reply:
This has been resolved, see updated PR